packages/playwright-core/src/server/screenshotter.ts

```diff
@@ -62,79 +62,50 @@ export class Screenshotter {
   async screenshotPage(progress: Progress, options: types.ScreenshotOptions): Promise<Buffer> {
     const format = validateScreenshotOptions(options);
     return this._queue.postTask(async () => {
-      const { viewportSize, originalViewportSize } = await this._originalViewportSize(progress);
+      const { viewportSize } = await this._originalViewportSize(progress);
 
+      const fullPageSize = await this._fullPageSize(progress);
+      let documentRect = { x: 0, y: 0, width: fullPageSize.width, height: fullPageSize.height };
+      const fitsViewport = fullPageSize.width <= viewportSize.width && fullPageSize.height <= viewportSize.height;
       if (options.fullPage) {
-        const fullPageSize = await this._fullPageSize(progress);
-        let documentRect = { x: 0, y: 0, width: fullPageSize.width, height: fullPageSize.height };
-        let overriddenViewportSize: types.Size | null = null;
-        const fitsViewport = fullPageSize.width <= viewportSize.width && fullPageSize.height <= viewportSize.height;
-        if (!this._page._delegate.canScreenshotOutsideViewport() && !fitsViewport) {
-          overriddenViewportSize = fullPageSize;
-          progress.throwIfAborted(); // Avoid side effects.
-          await this._page.setViewportSize(overriddenViewportSize);
-          progress.cleanupWhenAborted(() => this._restoreViewport(originalViewportSize));
-        }
         if (options.clip)
           documentRect = trimClipToSize(options.clip, documentRect);
-        const buffer = await this._screenshot(progress, format, documentRect, undefined, options);
+        const buffer = await this._screenshot(progress, format, documentRect, undefined, fitsViewport, options);
         progress.throwIfAborted(); // Avoid restoring after failure - should be done by cleanup.
-        if (overriddenViewportSize)
-          await this._restoreViewport(originalViewportSize);
         return buffer;
       }
 
       const viewportRect = options.clip ? trimClipToSize(options.clip, viewportSize) : { x: 0, y: 0, ...viewportSize };
-      return await this._screenshot(progress, format, undefined, viewportRect, options);
+      return await this._screenshot(progress, format, undefined, viewportRect, fitsViewport, options);
```

Comment:
`fitsViewport` should be always `true` here, no matter what full page size is. We are screenshotting the viewport, so we trim to clip and to viewport size in the previous line.